### ISSUE TITLE
Make it possible to compile a model from a package using the command line

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#10564]
### Description

I have a package laid out in a directory structure, and I would like to instantiate or run a model contained in it directly from the command line, without using a .mos script. Unfortunately, this is currently not possible: the omc command line structure only allows to instantiate or run a single **model** contained in a .mo file.

### Steps to Reproduce

Consider for example the attached MWE package: [FooBar.zip](https://github.com/OpenModelica/OpenModelica/files/11253432/FooBar.zip). If I type
```
myprompt> omc -i=FooBar.SubPackage.M FooBar/package.mo
```
I would expect OMC to load the whole FooBar package, and then to show me the flat code of model `FooBar.SubPackage.M`. What I get instead is:
```
Error processing file: FooBar/package.mo
Error: Class FooBar.SubPackage.M not found in scope <top>.
```
which basically means the package was not loaded.

### Expected Behavior

The documentation of omc from the command line should read:
```
Usage: omc [Options] (Class.mo | Script.mos) [Libraries | .mo-files] 

Loads Class.mo from the file system or runs the Script.mos script, loading other installed Libraries or other .mo files first. By default the output is the the flat model of Class.mo, but this can be changed by setting the appropriate Options.

* Libraries: Fully qualified names of libraries to load before processing Model or Script.
             The libraries should be separated by spaces: Lib1 Lib2 ... LibN.

...
* Examples:
  omc Model.mo             will produce flattened Model on standard output.
  omc -s Model.mo          will produce simulation code for the model:
                            * Model.c           The model C code.
                            * Model_functions.c The model functions C code.
                            * Model.makefile    The makefile to compile the model.
                            * Model_init.xml    The initial values.
  omc Script.mos           will run the commands from Script.mos.
  omc Model.mo Modelica    will first load the Modelica library and then produce 
                            flattened Model on standard output.
  omc Model1.mo Model2.mo  will load both Model1.mo and Model2.mo, and produce 
                            flattened Model1 on standard output.
  omc -i=MyLib.MyModel MyLib/package.mo will load the package MyLib and produce
            flattened MyLib.MyModel on standard output.
  *.mo (Modelica files) 
  *.mos (Modelica Script files)
```
and then the command line invocation should actually do it.

**Comment by mahge** [comment]
I looked at this and it might be possible to do. However, we need to define the behavior clearly using rules that we already have. I guess you probably want the same behavior you see in `OMEdit`.

----------------------
```sh
myprompt> omc -i=FooBar.SubPackage.M FooBar/package.mo
```

Do we want this to behave the same as this:

```sh
myprompt> omc -i=FooBar.SubPackage.M
```

IF we assume `FooBar` package was in `MODELICAPATH`?

---------------------------------------------------------------------------------

In other words, if we are asked to load a file named `package.mo` on the command-line, we will instead take the parent directory (`FooBar/`) and load it as a library instead. To do this we can temporarily add the parent-parent directory (`FooBar/../`) of the `package.mo` file to the MODELICAPATH and ask `omc` to treat the command as if it was 

```sh
myprompt> omc -i=FooBar.SubPackage.M FooBar
```

**Comment by mahge** [comment]
Another option which might be better is to **always look for libraries in the current directory** as well. That is MODELICAPATH will implicitly include the current directory as well. With this, your original command will be just:

```sh
myprompt> omc -i=FooBar.SubPackage.M 
```

and `omc` will try to load `FooBar` (using the same mechanism used to automatically load Modelica for example), but now with the current directory being considered as part of `MODELICAPATH`.

This is similar to what compilers like GCC/Clang perform for C/C++ include files.

**Comment by perost** [comment]
In my opinion loading a file on the command line and via `loadFile` should behave the same, so that `omc -i=FooBar.SubPackage.M FooBar/package.mo` would be equivalent to:
```
loadFile("FooBar/package.mo");
instantiateModel(FooBar.SubPackage.M);
```
This will currently load the whole package, while adding `FooBar/package.mo` on the command line will instead only load that file and nothing else.

**Comment by casella** [comment]
I agree, only loading `package.mo` doesn't make sense.

**Comment by mahge** [comment]
> In my opinion loading a file on the command line and via `loadFile` should behave the same, so that `omc -i=FooBar.SubPackage.M FooBar/package.mo` would be equivalent to:

Sounds good. As long as we define these features in terms of each other, it will remain sane. Should be fixed with #10748.

**Comment by mahge** [comment]
As it turns out `loadFile()` was not complaining on this model
```
model M
end M;

model M
end M;
```

On the other hand, loading it from the `commandline` issued an error. Now that `commandline` usage behavior is made the same as `loadFile()`, it too does not complain about this.

This can be fixed (made to issue an error) by changing this line https://github.com/OpenModelica/OpenModelica/blob/96fe37fc6082e3e14c71be20ae598e076275dac3/OMCompiler/Compiler/Script/CevalScript.mo#L354

to 
```diff
- p := InteractiveUtil.updateProgram(newp, p);
+ p := InteractiveUtil.mergeProgram(newp, p);
```

`InteractiveUtil.updateProgram()` will take the newly loaded class and then updates the AST replacing anything with the same name. `InteractiveUtil.mergeProgram()` will issue an error on duplicates. 

I am not sure what implications this will have on interactive usages. It is not possible to test all the different ways users load and use libraries. We can try it out, and if something goes wrong, we can revert it and look for an alternative.

**Comment by mahge** [comment]
@adrpo

**Comment by mahge** [comment]
> It might cause issues if loadFile is used to overwrite a previously loaded class

This is what I suspect as well. The test-suite is [running](https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-10748/2/pipeline) with the change to `mergeProgram` and we will see what it reports.  However, even if all tests pass, that is still not a full validation because clients like OMEdit might have their own expectations.

> But implementing that efficiently might be hard.

Exactly. It might be duplicate work to some extent. However, considering that this will only check duplicate classes at TOP level, it might not be so bad (to get the same behavior as before). Most files will have only one TOP level class. Even if not, it is not usually more than a handful.  So the total number of top level classes is close to the number of files we load. The check will probably be relatively low overhead compared to the actual parsing and loading of the file itself.

**Comment by casella** [comment]
> It might cause issues if `loadFile` is used to overwrite a previously loaded class. I think we might have some test cases that does that, but I can't say for sure. Ideally I think we should allow overwriting previously loaded classes, but issue an error if a file/library contains multiple definitions of the same name. But implementing that efficiently might be hard.

Do we have an unloadFile() API? If not, we need loadFile to overwrite previously loaded classes. Otherwise you can't reload updated versions from the file system, and you would need to quit OMC to do that.

**Comment by perost** [comment]
> > It might cause issues if `loadFile` is used to overwrite a previously loaded class. I think we might have some test cases that does that, but I can't say for sure. Ideally I think we should allow overwriting previously loaded classes, but issue an error if a file/library contains multiple definitions of the same name. But implementing that efficiently might be hard.
> 
> Do we have an unloadFile() API? If not, we need loadFile to overwrite previously loaded classes. Otherwise you can't reload updated versions from the file system, and you would need to quit OMC to do that.

We have `reloadClass` to reload a named class from the file associated with it, though it looks like it reloads the whole file and not just the given class. And there also the undocumented `deleteClass` which unloads a named class.

**Comment by casella** [comment]
Why is `deleteClas`s undocumented?

**Comment by perost** [comment]
> Why is `deleteClass` undocumented?

Because it's not documented? We have two types of API calls: those listed in `ModelicaBuiltin.mo` and implemented in `CevalScriptBackend.mo`, and those implemented in `Interactive.mo`. They work subtly different, and only the ones that are listed in `ModelicaBuiltin.mo` are actually documented. We've been slowly migrating API calls from `Interactive` to `CevalScriptBackend` over the years, but there are still a bunch left (see [this function](https://github.com/OpenModelica/OpenModelica/blob/71367fbce2ebfaa24c5f4242a5301a0a03a3fcbd/OMCompiler/Compiler/Script/Interactive.mo#L912)).

**Comment by casella** [comment]
OK, maybe we should have a plan for that.

**Comment by mahge** [comment]
#10748 is now updated to report duplicate top level classes. It does this on a file by file basis. So it will only report duplicates within a single file and not across multiple files. The performance should not be so bad since it will only do the extra check if the file happens to have more than one top level class. I think this is a very rare occurrence when libraries are considered.

I also had to adjust the error printing a bit. This means all messages (errors/warning/notifications) that use the `omc` Error handling mechanism will be printed after the flattened model is printed. I nice bonus side effect I think.

**Comment by bilderbuchi** [comment]
Just wanted to say that in the context of trying to set up testing for a local package, I just ran into the same issue. 
Thank you for fixing it, improved behaviour for `omc` is quite useful!
